### PR TITLE
One possible solution to add flexibility for mkldnn placement pass

### PR DESCRIPTION
**Pull request by bingyanghuang**
Solution for issue #14612 
This PR proposed one solution to selectively turn on "use_mkldnn" attribute in mkldnn_placement_pass. 
After investigation, we found that we can add attribute to each pass. So we decided to pass a string vector named op_list which contains all the op names we want to enable by this attribute of pass. 
You can add the following code to test this functionality in line 177.  
``` C++
 if (use_mkldnn) {    
      cfg.EnableMKLDNN();  
      std::unordered_set<std::string> op_list ={"conv3d"};  
      cfg.SetMKLDNNOp(op_list);  
 }
```
https://github.com/PaddlePaddle/Paddle/blob/e90afec47b0498e03950369323e4f85335c406f5/paddle/fluid/inference/tests/api/analyzer_dam_tester.cc#L177

This test will enable just conv3d  mkldnn kernel in DAM model. Meanwhile, you can test it with resnet50 to make sure that this works well with resnet50 too.

### CODE REVIEW

File: paddle/fluid/inference/api/analysis_predictor.cc [hunk 1]
```diff
@@ -324,7 +324,10 @@ void AnalysisPredictor::OptimizeInferenceProgram() {
     argument_.SetUseTensorRT(true);
     argument_.SetTensorRtWorkspaceSize(config_.tensorrt_workspace_size_);
     argument_.SetTensorRtMaxBatchSize(config_.tensorrt_max_batchsize_);
+  }else if(config_.use_mkldnn_){   
+    argument_.SetMKLDNNOp(config_.mkldnn_op_);
   }
```
Comment (luotao1) [inline paddle/fluid/inference/api/analysis_predictor.cc:328]: 327-328 is not related with tensorrt
```
if(config_.use_mkldnn_) {
  argument_.SetMKLDNNOp(config_.mkldnn_op_);
}
```

File: paddle/fluid/framework/ir/mkldnn_placement_pass.cc [hunk 2]
```diff
@@ -18,12 +18,22 @@ namespace paddle {
...
   VLOG(3) << "Aplies MKL-DNN placement strategy.";
+  const std::vector<std::string> op_list = 
+      Get<std::vector<std::string>>("mkldnn_op_list");
   for (const Node* n : graph->Nodes()) {
     if (n->IsOp() && n->RuntimeHasAttr("use_mkldnn")) {
-      n->Op()->SetAttr("use_mkldnn", true);
+      if (op_list.empty()) {
+        n->Op()->SetAttr("use_mkldnn", true);
+      } else {
+          if (std::find(op_list.begin(), op_list.end(), n->Name()) !=
```
Comment (luotao1) [inline paddle/fluid/framework/ir/mkldnn_placement_pass.cc]: `else if (std::find(op_list.begin(), op_list.end(), n->Name())`

File: paddle/fluid/framework/ir/mkldnn_placement_pass.cc [hunk 3]
```diff
@@ -18,12 +18,22 @@ namespace paddle {
 namespace framework {
 namespace ir {
 
+
 std::unique_ptr<ir::Graph> MKLDNNPlacementPass::ApplyImpl(
     std::unique_ptr<ir::Graph> graph) const {
   VLOG(3) << "Aplies MKL-DNN placement strategy.";
+  const std::vector<std::string> op_list =
```
Comment (Superjomn) [inline paddle/fluid/framework/ir/mkldnn_placement_pass.cc]: `const auto& op_list = Get<std::vector<std::string>>("mkldnn_op_list");`

File: paddle/fluid/inference/analysis/argument.h [hunk 4]
```diff
@@ -114,6 +114,7 @@ struct Argument {
   // The ir passes to perform in analysis phase.
   DECL_ARGUMENT_FIELD(ir_analysis_passes, IrAnalysisPasses,
                       std::vector<std::string>);
+  DECL_ARGUMENT_FIELD(mkldnn_op, MKLDNNOp, std::vector<std::string>); //Choose to enable mkldnn op
```
Comment (Superjomn) [inline paddle/fluid/inference/analysis/argument.h]: mkdnn_enabled_op_types

File: paddle/fluid/framework/ir/mkldnn_placement_pass.cc [hunk 5]
```diff
@@ -34,4 +44,5 @@ std::unique_ptr<ir::Graph> MKLDNNPlacementPass::ApplyImpl(
 }  // namespace paddle
 
 REGISTER_PASS(mkldnn_placement_pass,
-              paddle::framework::ir::MKLDNNPlacementPass);
+              paddle::framework::ir::MKLDNNPlacementPass)
+     .RequirePassAttr("mkldnn_op_list");
```
Comment (Superjomn) [inline paddle/fluid/framework/ir/mkldnn_placement_pass.cc]: mkldnn_op_list -> mkldnn_enabled_op_types

the op list is just a list of op types, and better to be a `unordered_set`.

File: paddle/fluid/framework/ir/mkldnn_placement_pass.cc (hunk 2 above)
Comment (Superjomn) [inline paddle/fluid/framework/ir/mkldnn_placement_pass.cc]: the `op_list` is better to be a `set`, not a vector.

File: paddle/fluid/inference/analysis/argument.h (hunk 4 above)
Comment (Superjomn) [reply paddle/fluid/inference/analysis/argument.h]: add some comments here to describe this argument.

File: paddle/fluid/inference/analysis/ir_pass_manager.cc [hunk 6]
```diff
@@ -63,6 +63,9 @@ void IRPassManager::CreatePasses(Argument *argument,
       pass->Set("graph_viz_path", new std::string(std::move(dot_file_path)));
       pass_num++;
     }
+    if (pass_name == "mkldnn_placement_pass"){
+       pass->Set("mkldnn_op_list", new std::vector<std::string>(argument->mkldnn_op()));
+    }
```
Comment (Superjomn) [inline paddle/fluid/inference/analysis/ir_pass_manager.cc:70]: `pass->Set("mkldnn_enabled_op_types", new std::vector<std::string>(argument->mkldnn_enabled_op_types()));`

File: paddle/fluid/inference/api/paddle_analysis_config.h [hunk 7]
```diff
@@ -55,12 +55,14 @@ struct AnalysisConfig : public NativeConfig {
...
   // NOT stable yet.
   bool use_mkldnn() const { return use_mkldnn_; }
+  void SetMKLDNNOp(std::vector<std::string> op_list){ mkldnn_op_ = op_list;}
 
   friend class ::paddle::AnalysisPredictor;
 
  protected:
   bool use_tensorrt_{false};
   bool use_mkldnn_{false};
+  std::vector<std::string> mkldnn_op_;
```
Comment (Superjomn) [inline paddle/fluid/inference/api/paddle_analysis_config.h]: vector -> set

**Review by Superjomn — CHANGES_REQUESTED** [review]

File: paddle/fluid/inference/api/analysis_predictor.cc (hunk 1 above)
Comment (bingyanghuang) [reply paddle/fluid/inference/api/analysis_predictor.cc:328]: Will change it , thanks

File: paddle/fluid/framework/ir/mkldnn_placement_pass.cc (hunk 2 above)
Comment (bingyanghuang) [reply paddle/fluid/framework/ir/mkldnn_placement_pass.cc]: Will change it , thanks

File: paddle/fluid/framework/ir/mkldnn_placement_pass.cc (hunk 3 above)
Comment (bingyanghuang) [reply paddle/fluid/framework/ir/mkldnn_placement_pass.cc]: Will change it , thanks

File: paddle/fluid/inference/analysis/argument.h (hunk 4 above)
Comment (bingyanghuang) [reply paddle/fluid/inference/analysis/argument.h]: Got it, will change it and add some comments. Thanks.

File: paddle/fluid/framework/ir/mkldnn_placement_pass.cc (hunk 5 above)
Comment (bingyanghuang) [reply paddle/fluid/framework/ir/mkldnn_placement_pass.cc]: Has changed, thanks.

File: paddle/fluid/framework/ir/mkldnn_placement_pass.cc (hunk 2 above)
Comment (bingyanghuang) [reply paddle/fluid/framework/ir/mkldnn_placement_pass.cc]: Got it.

File: paddle/fluid/inference/analysis/ir_pass_manager.cc (hunk 6 above)
Comment (bingyanghuang) [reply paddle/fluid/inference/analysis/ir_pass_manager.cc:70]: Done.

File: paddle/fluid/inference/api/paddle_analysis_config.h (hunk 7 above)
Comment (bingyanghuang) [reply paddle/fluid/inference/api/paddle_analysis_config.h]: Done.

**Comment by luotao1** [comment]
please fix ci, and add following two lines in test:
```
std::unordered_set<std::string> op_list ={"conv3d"};  
cfg.SetMKLDNNOp(op_list);  
```
https://github.com/PaddlePaddle/Paddle/blob/82eefceabedfbed60d9edb6ec219cfc4920d27aa/paddle/fluid/inference/tests/api/analyzer_dam_tester.cc#L195-L197

**Comment by bingyanghuang** [comment]
> please fix ci, and add following two lines in test:
> 
> ```
> std::unordered_set<std::string> op_list ={"conv3d"};  
> cfg.SetMKLDNNOp(op_list);  
> ```
> [Paddle/paddle/fluid/inference/tests/api/analyzer_dam_tester.cc](https://github.com/PaddlePaddle/Paddle/blob/82eefceabedfbed60d9edb6ec219cfc4920d27aa/paddle/fluid/inference/tests/api/analyzer_dam_tester.cc#L195-L197)
> 
> Lines 195 to 197 in [82eefce](/PaddlePaddle/Paddle/commit/82eefceabedfbed60d9edb6ec219cfc4920d27aa)
> 
>  if (use_mkldnn) { 
>    cfg.EnableMKLDNN(); 
>  }

Got it. Will change right now.

File: paddle/fluid/inference/tests/api/analyzer_dam_tester.cc [hunk 8]
```diff
@@ -194,6 +194,8 @@ void profile(bool use_mkldnn = false) {
 
   if (use_mkldnn) {
     cfg.EnableMKLDNN();
+    std::unordered_set<std::string> op_list = {"conv3d"};
+    cfg.SetMKLDNNOp(op_list);
```
Comment (luotao1) [inline paddle/fluid/inference/tests/api/analyzer_dam_tester.cc:198]: please update the compare funcation as well.
```
void compare(bool use_mkldnn = false) {
  AnalysisConfig cfg;
  SetConfig(&cfg);
  if (use_mkldnn) {
    cfg.EnableMKLDNN();
  }
```

Comment (bingyanghuang) [reply paddle/fluid/inference/tests/api/analyzer_dam_tester.cc:198]: Done, Thanks for your advice.

**Review by luotao1 — APPROVED** [review]
LGTM！

**Review by Superjomn — APPROVED** [review]
LGTM

**Pull request merged**